src/app/icd/client/CheckInHandler.cpp

```diff
@@ -109,8 +109,13 @@ CHIP_ERROR CheckInHandler::OnMessageReceived(Messaging::ExchangeContext * ec, co
 
     if (refreshKey)
     {
-        // TODO: A new CASE session should be established to re-register the client using a new key. The registration will happen in
-        // CASE session callback
+        ICDRefreshKeyInfo * refreshKeyInfo = mpCheckInDelegate->OnKeyRefreshNeeded(clientInfo, mpICDClientStorage);
+        if (refreshKeyInfo == nullptr)
+        {
+            ChipLogError(ICD, "Key Refresh failed");
+            return CHIP_NO_ERROR;
+        }
+        refreshKeyInfo->EstablishSessionToPeer();
```

Comment:
Doesn't this leak on failure?

Reply:
All failures after ICDRefreshKeyInfo is created will call OnRegistrationUpdateFailure. I'll add them.